sound/soc/sof/trace.c

```diff
@@ -53,73 +50,64 @@ static int sof_wait_trace_avail(struct snd_sof_dev *sdev, size_t *count,
 
 	if (signal_pending(current)) {
 		remove_wait_queue(&sdev->trace_sleep, &wait);
-		goto _endcheck;
+		goto out;
 	}
 
 	/* set timeout to max value, no error code */
 	schedule_timeout(MAX_SCHEDULE_TIMEOUT);
 	remove_wait_queue(&sdev->trace_sleep, &wait);
 
-_endcheck:
-	/* calculate the available count */
-	avail = sdev->host_offset - pos;
-
-_host_end:
-	/* return min value between available and request count */
-	*count = avail < *count ? avail : *count;
-
-	return 0;
+out:
+	/* return bytes available for copy */
+	if (sdev->host_offset < pos)
+		return buffer_size - pos;
+	else
+		return sdev->host_offset - pos;
 }
 
 static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
 				       size_t count, loff_t *ppos)
 {
 	struct snd_sof_dfsentry_buf *dfse = file->private_data;
 	struct snd_sof_dev *sdev = dfse->sdev;
-	int err;
-	loff_t pos = *ppos;
-	loff_t lpos = pos;
-	size_t ret, size;
+	unsigned long rem;
+	loff_t lpos = *ppos;
+	size_t avail, buffer_size = dfse->size;
 
-	size = dfse->size;
+	/* make sure we know about any failures on the DSP side */
+	sdev->dtrace_error = false;
 
 	/* check pos and count */
-	if (pos < 0)
+	if (lpos < 0)
 		return -EINVAL;
 	if (!count)
 		return 0;
 
-	/*
-	 * If pos exceeds size, it means host DMA buffer has been wrapped. So
-	 * local pos will be truncated from global pos. It is possible to wrap
-	 * host DMA buffer multiply times when keep output long time, so we
-	 * need one loop to process it.
-	 */
-	while (lpos >= size)
-		lpos -= size;
-
-	if (count > size - lpos)
-		count = size - lpos;
+	/* check for buffer wrap and count overflow */
+	lpos = lpos % buffer_size;
+	if (count > buffer_size - lpos)
+		count = buffer_size - lpos;
 
 	/* get available count based on current host offset */
-	err = sof_wait_trace_avail(sdev, &count, lpos, size);
-	if (err < 0) {
-		dev_err(sdev->dev,
-			"error: can't get more trace %d\n", err);
-		return 0;
+	avail = sof_wait_trace_avail(sdev, lpos, buffer_size);
+	if (sdev->dtrace_error) {
+		dev_err(sdev->dev, "error: trace IO error\n");
+		return -EIO;
 	}
 
-	/* copy available trace data to debugfs */
-	ret = copy_to_user(buffer, dfse->buf + lpos, count);
+	/* make sure count is <= avail */
+	count = avail > count ? count : avail;
 
-	if (ret == count)
+	/* copy available trace data to debugfs */
+	rem = copy_to_user(buffer, dfse->buf + lpos, count);
+	if (rem == count)
 		return -EFAULT;
-	count -= ret;
 
-	/* move debugfs reading position */
-	*ppos = pos + count;
+	count -= rem;
+	*ppos += count;
 
-	return count;
+	/* move debugfs reading position */
+	return *ppos;
```

Comment:
not sure I get the change here, why not return count as before?

Reply:
oops, backport merge error.